proof/refine/ARM/Ipc_R.thy

```diff
@@ -4946,9 +4940,96 @@ lemma valid_replies'_no_tcb_not_linked:
   apply (case_tac a; simp)
   done
 
-lemma replyPush_sym_refs_list_refs_of_replies'[wp]:
-  "\<lbrace>(\<lambda>s. sym_refs (list_refs_of_replies' s)) and valid_replies'
-    and obj_at' (\<lambda>ko. replyTCB ko = None) replyPtr\<rbrace>
+(* START -- can these be moved higher? *)
+
+lemma updateReply_reply_at'_wp:
+  "\<lbrace>\<lambda>s. P (if p = rptr then True else reply_at' p s)\<rbrace>
+   updateReply rptr f
+   \<lbrace>\<lambda>rv s. P (reply_at' p s)\<rbrace>"
+  apply (rule hoare_weaken_pre, rule updateReply_obj_at')
+  apply (clarsimp simp: obj_at'_real_def split: if_splits)
+  done
+
+crunches setThreadState
+  for tcb_scs_of'[wp]: "\<lambda>s. P (tcb_scs_of' s)"
+  (ignore: threadSet wp: threadSet_tcb_scs_of'_inv)
+
+crunches replyUnlink
+  for tcb_scs_of'[wp]: "\<lambda>s. P (tcb_scs_of' s)"
+  and scs_of'[wp]: "\<lambda>s. P (scs_of' s)"
+
+lemma replyUnlink_tcbs_scs_of[wp]:
+  "replyUnlink rPtr tPtr \<lbrace>\<lambda>s. P (tcb_scs_of' s) (scTCBs_of s)\<rbrace>"
+  by (rule hoare_weaken_pre, wps, wp, simp)
+
+lemma replyUnlink_replies_scs_of[wp]:
+  "replyUnlink rPtr tPtr \<lbrace>\<lambda>s. P (scReplies_of s) (replySCs_of s)\<rbrace>"
+  by (rule hoare_weaken_pre, wps, wp, simp)
+
+lemma schedContextUpdateConsumed_scReplies_of[wp]:
+  "schedContextUpdateConsumed scPtr \<lbrace>\<lambda>s. P (scReplies_of s) \<rbrace>"
+  unfolding schedContextUpdateConsumed_def
+  apply (wpsimp simp: setSchedContext_def)
+  apply (clarsimp simp: opt_map_def if_distrib)
+  apply (erule subst[where P=P, rotated], rule ext)
+  apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def split: option.splits)
+  done
+
+lemma schedContextUpdateConsumed_sc_tcbs_of[wp]:
+  "schedContextUpdateConsumed scPtr \<lbrace>\<lambda>s. P (scTCBs_of s)\<rbrace>"
+  unfolding schedContextUpdateConsumed_def
+  apply (wpsimp simp: setSchedContext_def)
+  apply (clarsimp simp: opt_map_def if_distrib)
+  apply (erule subst[where P=P, rotated], rule ext)
+  apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def split: option.splits)
+  done
+
+crunches schedContextUpdateConsumed
+  for tcb_scs_of'[wp]: "\<lambda>s. P (tcb_scs_of' s)"
+
+crunches doIPCTransfer
+  for replySCs_of[wp]: "\<lambda>s. P (replySCs_of s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma schedContextUpdateConsumed_misc_heaps[wp]:
+  "schedContextUpdateConsumed scPtr \<lbrace>\<lambda>s. P (scReplies_of s) (replySCs_of s)\<rbrace>"
+  "schedContextUpdateConsumed scPtr \<lbrace>\<lambda>s. P (tcb_scs_of' s) (scTCBs_of s)\<rbrace>"
```

Comment:
Wouldn't it be slightly stronger to combine these into one condition that handles all of the heaps? I don't have anything in mind that it would actually help with, but I also don't see a benefit in splitting them apart.

Reply:
Great point!